src/client/types.ts

```diff
@@ -1,14 +1,18 @@
-import type * as grpc from '@grpc/grpc-js';
-import type { Class } from '@matrixai/errors';
-import type ErrorPolykey from '../ErrorPolykey';
+import type { JSONValue } from '../types';
 
-type Authenticate = (
-  metadataClient: grpc.Metadata,
-  metadataServer?: grpc.Metadata,
-) => Promise<grpc.Metadata>;
+// eslint-disable-next-line
+type NoData = {};
 
-type ClientClientErrors = Array<
-  Class<ErrorPolykey<any>> | Array<Class<ErrorPolykey<any>>>
->;
+type RPCRequestParams<T extends Record<string, JSONValue> = NoData> = {
+  metadata?: {
+    [Key: string]: JSONValue;
+  } & Partial<{ Authorization: string }>;
+} & Omit<T, 'metadata'>;
 
-export type { Authenticate, ClientClientErrors };
+type RPCResponseResult<T extends Record<string, JSONValue> = NoData> = {
+  metadata?: {
+    [Key: string]: JSONValue;
+  } & Partial<{ Authorization: string }>;
+} & Omit<T, 'metadata'>;
+
+export type { RPCRequestParams, RPCResponseResult, NoData };
```

Comment:
Aren't these generic to `rpc`? Why not have these types in `rpc`?

Reply:
Yes and no? By design the RPC doesn't care about the parameter structure. This is here to enforce the metadata structure for the `client` domain to be used by the middleware, handler and client implementation.

So it's not used at all in the context of the `rpc` domain, but will be used by both the `client` and `agent` domains, they'll share the middleware code as well. So with that maybe this should be moved to a common place between the two? Not sure.

Reply:
If they are specific to client/agent "services", one should make that clearer by saying `ClientRPCRequestParams` or `AgentRPCRequestParams`. Right now these look like `rpc` domain types.

Reply:
They're getting pretty verbose already. Maybe change them to `ClientRPCReq`, `ClientRPCRes`, `AgentRPCReq` and `AgentRPCRes`?

Reply:
I don't like shortening type names. Type names can verbose, it's fine.